### PR TITLE
Check if user exists when accepting registration requests

### LINKED ISSUE #6651: Accepting an account request with an email that is already in use triggers an exception
We had a situation where a user had a pending request for a local account, then logged in through another authentication provider and changed his email address to that of the request. This triggered an exception when I tried to accept the request.

Below are the steps of what happened. It would be nice if the conflict could be detected before the exception is thrown.

This is in Indico v3.2.9.

1. A user (who also has an LDAP account) requests a new Indico profile with his `cern.ch` email:
```text
2024-12-02 18:48:52,513  INFO     19b6754216884fe9  indico.rh                 GET /register/ [IP=134.79.138.4] [PID=50292]
2024-12-02 18:48:52,548  INFO     0000000000000000  indico.emails             Sent email "[Indico] Verify your email"
2024-12-02 18:48:58,384  INFO     c363e5cd5c704caf  indico.rh                 GET /register/?next=%2Fevent%2F8288%2Fcontributions%2F7762%2Fauthor%2F11559 [IP=134.79.138.4] [PID=50292]
<...>
2024-12-02 18:49:41,070  INFO     0000000000000000  indico.emails             Sent email "[Indico] Indico profile requested"
```
Email:
```text
Subject: [Indico] Indico profile requested
Date: Tue, 03 Dec 2024 02:49:41 -0000
From: noreply@slac.stanford.edu
<...>
First name: Bart
Last name: Simpson
Email address: bart.simpson@cern.ch
Affiliation: University of California Santa Cruz
```

2. The user reconsiders and logs in through LDAP (for the first time):
```text
2024-12-02 18:49:50,751  INFO     b2649396af0249cc  indico.auth               Received identity info: <IdentityInfo(<LDAPIdentityProvider(ldap)>, bsimpson, address=None, affiliation='SLAC', affiliation_data=None, email=None, first_name='Bart', last_name='Simpson', phone=None)>
2024-12-02 18:49:50,784  INFO     b2649396af0249cc  indico.auth               Found existing identity <Identity(1730, 1858, ldap, bsimpson)> for user <User(1858, bsimpson@slac.stanford.edu): "Bart Simpson">
2024-12-02 18:49:50,784  INFO     b2649396af0249cc  indico.auth               Updated data of identity <Identity(1730, 1858, ldap, bsimpson)> for user <User(1858, bsimpson@slac.stanford.edu): "Bart Simpson">
```

3. The user updates his affiliation and primary email address to that of the request:
```text
2024-12-02 21:37:57,994  INFO     d4a82921b5fc4440  indico.users              Profile of user <User(1858, bart.simpson@cern.ch): "Bart Simpson"> updated by <User(1858, bart.simpson@cern.ch): "Bart Simpson">: {'affiliation': 'University of California, Santa Cruz', 'affiliation
_link': None}
```

4. The admin attempts to accept the pending registration and triggers an exception:
```text
2024-12-03 09:12:23,322  INFO     8450dcc201f347c0  indico.rh                 POST /admin/users/registration-requests/1575/accept [IP=134.79.138.4] [PID=63699]
2024-12-03 09:12:23,358  ERROR    8450dcc201f347c0  indico.flask              (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "ix_uq_emails_email"

[SQL: INSERT INTO users.emails (user_id, email, is_primary, is_user_deleted) VALUES (%(user_id)s, %(email)s, %(is_primary)s, %(is_user_deleted)s) RETURNING users.emails.id]
[parameters: {'user_id': 3216, 'email': 'bart.simpson@cern.ch', 'is_primary': True, 'is_user_deleted': False}]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
Traceback (most recent call last):
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1905, in _execute_context
    self.dialect.do_execute(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "ix_uq_emails_email"
DETAIL:  Key (email)=(bart.simpson@cern.ch) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/indico/.venv/lib/python3.9/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/indico/.venv/lib/python3.9/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/web/flask/util.py", line 79, in wrapper
    return obj().process()
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/web/rh.py", line 295, in process
    handle_sqlalchemy_database_error()  # this will re-raise an exception
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/web/rh.py", line 285, in process
    res = self._do_process()
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/web/rh.py", line 256, in _do_process
    rv = self._process()
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/modules/users/controllers.py", line 751, in _process
    user, identity = register_user(self.request.email, self.request.extra_emails, self.request.user_data,
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/modules/auth/util.py", line 57, in register_user
    user = create_user(email, user_data, identity=identity, settings=settings, other_emails=extra_emails,
  File "/opt/indico/.venv/lib/python3.9/site-packages/indico/modules/users/operations.py", line 56, in create_user
    db.session.flush()
  File "<string>", line 2, in flush
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3449, in flush
    self._flush(objects)
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3589, in _flush
    transaction.rollback(_capture_exception=True)
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3549, in _flush
    flush_context.execute()
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    _emit_insert_statements(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1238, in _emit_insert_statements
    result = connection._execute_20(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1710, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement
    ret = self._execute_context(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1948, in _execute_context
    self._handle_dbapi_exception(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2129, in _handle_dbapi_exception
    util.raise_(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1905, in _execute_context
    self.dialect.do_execute(
  File "/opt/indico/.venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "ix_uq_emails_email"
DETAIL:  Key (email)=(bart.simpson@cern.ch) already exists.

[SQL: INSERT INTO users.emails (user_id, email, is_primary, is_user_deleted) VALUES (%(user_id)s, %(email)s, %(is_primary)s, %(is_user_deleted)s) RETURNING users.emails.id]
[parameters: {'user_id': 3216, 'email': 'bart.simpson@cern.ch', 'is_primary': True, 'is_user_deleted': False}]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
```

### PR DESCRIPTION
**Pull request by tomasr8**
Closes #6651

How to test:
- Register a new account without accepting it
- Change the email (primary or secondary) of another existing account to be the same as the one used to register the new account
- Go to `/admin/users/registration-requests/` and accept the registration request
- You should see a message saying that the user already exists
- The registration request should be deleted afterwards

### CODE REVIEW

File: indico/modules/users/controllers.py [hunk 1]
```diff
@@ -852,6 +852,12 @@ class RHAcceptRegistrationRequest(RHRegistrationRequestBase):
     """Accept a registration request."""
 
     def _process(self):
+        user_exists = User.query.filter(User.all_emails == self.request.email).has_rows()
+        if user_exists:
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:856]:
```suggestion
        if get_user_by_email(self.request.email):
```

That way you get the check for `~User.is_deleted` for free (deleted users may have colliding emails).

(You'll probably need to import it as well, I didn't check if this file is already importing it or not)

Comment (tomasr8) [reply indico/modules/users/controllers.py:856]: Ah right, we don't enforce uniqueness of emails for deleted users.

**Review by AjobK — APPROVED** [review]
LGTM :+1:

**Review by ThiefMaster — APPROVED** [review]

**Review by GovernmentPlates — APPROVED** [review]

**Comment by bartoldu** [comment]
Thanks!

**Pull request merged**